sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.cu

```diff
@@ -0,0 +1,471 @@
+      mem_pool_, cudaMemPoolAttrReleaseThreshold, static_cast<void *>(&pool_release_threshold)));
+  }
+
+  thrust_custom_allocator_ = std::make_unique<ThrustCustomAllocator>(stream_);
+}
+
+std::unique_ptr<cuda_blackboard::CudaPointCloud2> CudaVoxelGridDownsampleFilter::filter(
+  const cuda_blackboard::CudaPointCloud2::ConstSharedPtr & input_points)
+{
+  voxel_info_.num_input_points = input_points->width * input_points->height;
+  voxel_info_.input_point_step = input_points->point_step;
+
+  auto get_offset = [&](const std::string & field_name) -> size_t {
+    int index = -1;
+    for (size_t i = 0; i < input_points->fields.size(); ++i) {
+      if (input_points->fields[i].name == field_name) {
+        index = static_cast<int>(i);
+      }
+    }
+    if (index < 0) {
+      std::stringstream ss;
+      ss << "input cloud does not contain filed named '" << field_name << "'";
+      throw std::runtime_error(ss.str());
+    }
+    return input_points->fields[index].offset;
+  };
+
+  voxel_info_.input_xyzi_offset[0] = get_offset("x");
+  voxel_info_.input_xyzi_offset[1] = get_offset("y");
+  voxel_info_.input_xyzi_offset[2] = get_offset("z");
+  voxel_info_.input_xyzi_offset[3] = get_offset("intensity");
+  try {
+    voxel_info_.input_return_type_offset.offset = get_offset("return_type");
+    voxel_info_.input_return_type_offset.is_valid = true;
+  } catch (const std::runtime_error & e) {
+    voxel_info_.input_return_type_offset.is_valid = false;
+  }
+
+  try {
+    voxel_info_.input_channel_offset.offset = get_offset("channel");
+    voxel_info_.input_channel_offset.is_valid = true;
+  } catch (const std::runtime_error & e) {
+    voxel_info_.input_channel_offset.is_valid = false;
+  }
+
+  // Pull working buffer from the pooled region
+  auto coord_buffer_dev = allocateBufferFromPool<float>(voxel_info_.num_input_points);
+  auto voxel_index_buffer_dev = allocateBufferFromPool<size_t>(voxel_info_.num_input_points);
+  auto point_index_buffer_dev = allocateBufferFromPool<size_t>(voxel_info_.num_input_points);
+  auto index_map_dev = allocateBufferFromPool<size_t>(voxel_info_.num_input_points);
+  auto return_type_field_dev = allocateBufferFromPool<decltype(OutputPointType::return_type)>(1);
+  auto channel_field_dev = allocateBufferFromPool<decltype(OutputPointType::channel)>(1);
+
+  // Get min/max value for x, y, z in input_points to determine computation target range
+  getVoxelMinMaxCoordinate(input_points, coord_buffer_dev);
+
+  // Calculate unique voxel indices that all points in input_points belong to
+  // Create lookup table to show correspondence between spatial voxel grid position
+  // and memory index
+  auto num_valid_voxel = searchValidVoxel(
+    input_points, voxel_index_buffer_dev, point_index_buffer_dev, return_type_field_dev,
+    channel_field_dev);
+
+  // Allocate region actually result filtered pointcloud requires and fill them
+  auto filtered_output = std::make_unique<cuda_blackboard::CudaPointCloud2>();
```

Comment:
nit: allocation can be performed after publishing by allocating space of the larger pointcloud processed so far (the amount of cuda memory is negligible I think)